xmppserver/src/main/java/org/jivesoftware/openfire/spi/ConnectionType.java

```diff
@@ -53,6 +53,15 @@
         this.fallback = fallback;
     }
 
+    /**
+     * Flag that indicates whether the connections of this type are client-oriented.
+     * @return true if it is SOCKET_C2S or its fallback is SOCKET_C2S.
+     */
+    public boolean isClientOriented()
+    {
+        return SOCKET_C2S == this || SOCKET_C2S == this.getFallback();
```

Comment:
It would be safer to recursively verify fallbacks, to catch instances where the fallback of the fallback matches.